### PR TITLE
Add ClusterfuzzLite

### PR DESCRIPTION
**Pull request by AdamKorcz**
Signed-off-by: AdamKorcz <Adam@adalogics.com>

## Change Summary

This PR adds [ClusterfuzzLite](https://google.github.io/clusterfuzzlite/) and a fuzzer for the parser. 

ClusterfuzzLite will run fuzzers in the CI when PRs are made. It can be extended beyond this PR with code coverage and batch fuzzing: https://google.github.io/clusterfuzzlite/running-clusterfuzzlite/github-actions/

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: .github/workflows/cflite.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      id: build
+      uses: google/clusterfuzzlite/actions/build_fuzzers@v1
+      with:
+        sanitizer: ${{ matrix.sanitizer }}
+        language: python
+    - name: Run Fuzzers (${{ matrix.sanitizer }})
+      id: run
+      uses: google/clusterfuzzlite/actions/run_fuzzers@v1
+      with:
+        github-token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (samuelcolvin) [inline .github/workflows/cflite.yaml:26]: why do we need the github token?

File: pydantic/fuzz_parse.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/python3
```
Comment (samuelcolvin) [inline pydantic/fuzz_parse.py:1]: why do we need this file? Surely the source shouldn't have to change.

**Review by samuelcolvin — COMMENTED** [review]
Thanks for the PR, this is interesting.

Perhaps it would be good to start by adding this to [pydantic-core](https://github.com/samuelcolvin/pydantic-core) which will be the basis for future versions of pydantic?

File: pydantic/fuzz_parse.py (hunk 2 above)
Comment (AdamKorcz) [reply pydantic/fuzz_parse.py]: This file is the fuzz harness/test itself. The fuzz engine - [Atheris](https://github.com/google/atheris) - will run this test repeatedly.

This PR does not alter the pydantic source code but adds a fuzz test. The fuzzer can be placed anywhere in this repository though.

File: .github/workflows/cflite.yaml (hunk 1 above)
Comment (AdamKorcz) [reply .github/workflows/cflite.yaml:26]: It is a requirement for ClusterfuzzLite to work.

File: pydantic/fuzz_parse.py (hunk 2 above)
Comment (samuelcolvin) [reply pydantic/fuzz_parse.py]: Okay, well it definitely shouldn't be in `pydantic`, best to put it in `tests`

File: .github/workflows/cflite.yaml (hunk 1 above)
Comment (samuelcolvin) [reply .github/workflows/cflite.yaml:26]: Well I assume that, but why?

**Comment by samuelcolvin** [comment]
@Zac-HD I would love your opinion on this?

What do you think of [ClusterfuzzLite](https://google.github.io/clusterfuzzlite/), will it provide any utility here?

**Comment by Zac-HD** [comment]
I think fuzzing can be really helpful for native code, and the Clusterfuzz / OSS-Fuzz team are great.

For Python code, you can repurpose any or all of your existing property-based tests as fuzz targets ([hypothesis docs](https://hypothesis.readthedocs.io/en/latest/details.html#use-with-external-fuzzers), [OSS-fuzz docs](https://google.github.io/oss-fuzz/getting-started/new-project-guide/python-lang/#hypothesis)) and get the benefit of Hypothesis-style replay, minimal counterexamples, explain-mode, etc.  (I'm pretty fond of [HypoFuzz](https://hypofuzz.com/) for Python projects, but don't yet have a great story for sharing the database between Actions-for-fuzzing and Actions-for-CI.  In the meantime, Clusterfuzz does handle that and offer free compute!)

So I'd try it out, ideally with a focus on compiled code, and you can always drop it later if it's not worth the hassle!

**Comment by samuelcolvin** [comment]
Thanks, fuzzing found all the errors in speedate, so I'm a complete convert for native code.

**Comment by samuelcolvin** [comment]
It looks like tests are failing while "building the fuzzer" because the environment is somehow set up wrong and it's trying to use `pydantic/typing.py` on `from typing import ...` commands:

```
Compiling libFuzzer to /usr/lib/libFuzzingEngine.a...  done.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/atheris/__init__.py", line 17, in <module>
    from typing import List
  File "/src/pydantic/typing.py", line 4, in <module>
    from typing import (  # type: ignore
ImportError: cannot import name 'TYPE_CHECKING'
```

@AdamKorcz any idea why this is happening? I guess the next question is to ask on https://github.com/google/atheris or https://github.com/google/clusterfuzzlite

**Comment by AdamKorcz** [comment]
> It looks like tests are failing while "building the fuzzer" because the environment is somehow set up wrong and it's trying to use `pydantic/typing.py` on `from typing import ...` commands:
> 
> ```
> Compiling libFuzzer to /usr/lib/libFuzzingEngine.a...  done.
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "/usr/local/lib/python3.8/site-packages/atheris/__init__.py", line 17, in <module>
>     from typing import List
>   File "/src/pydantic/typing.py", line 4, in <module>
>     from typing import (  # type: ignore
> ImportError: cannot import name 'TYPE_CHECKING'
> ```
> 
> @AdamKorcz any idea why this is happening? I guess the next question is to ask on https://github.com/google/atheris or https://github.com/google/clusterfuzzlite

Let me have a look to see if I can see the issue.

**Comment by samuelcolvin** [comment]
No updates and I fail to see how this approach could realistically find anything interesting in pydantic.

@AdamKorcz if you'd like to submit a PR to the [speedate}(https://github.com/pydantic/speedate) project which is also on this organisation, I think that could definitely be useful. In particular speedate does quite a lot of manual parsing and already has fuzz tests that ClusterfuzzLite could perhaps hook into.

**Pull request closed without merging**